common/autoware_testing/design/autoware_testing-design.md

```diff
@@ -1,65 +1,117 @@
 
 The package aims to provide a unified way to add standard testing functionality to the package, currently supporting:
+
 - Smoke testing (`add_smoke_test`): launch a node with default configuration and ensure that it starts up and does not crash.
+- Interface testing (`add_interface_test`): launch a node with default configuration and topic configuration and ensure that it publishes and subscribes to the expected topics.
 
 # Design
 
 Uses `ros_testing` (which is an extension of `launch_testing`) and provides some parametrized, reusable standard tests to run.
 
 ## Assumptions / Known limits
 
-Parametrization is limited to package, executable names, parameters filename and executable arguments. Test namespace is set as 'test'.
-Parameters file for the package is expected to be in `param` directory inside package.
+Parametrization is limited to package, executable names, parameters filename, topics filename and executable arguments. Test namespace is set as 'test'.
+Parameters file and topics file for the package is expected to be in `param` directory inside package.
 
 ## Inputs / Outputs / API
 
+### Smoke test
+
 To add a smoke test to your package tests, add test dependency on `autoware_testing` to `package.xml`
 
-```{xml}
+```xml
 <test_depend>autoware_testing</test_depend>
 ```
 
 and add the following two lines to `CMakeLists.txt` in the `IF (BUILD_TESTING)` section:
 
-```{cmake}
+```cmake
 find_package(autoware_testing REQUIRED)
 add_smoke_test(<package_name> <executable_name> [PARAM_FILENAME <param_filename>] [EXECUTABLE_ARGUMENTS <arguments>])
 ```
 
-Where 
+Where
 
 `<package_name>` - [required] tested node package name.
 
 `<executable_name>` - [required] tested node executable name.
 
-`<param_filename>` - [optional] param filename. Default value is `test.param.yaml`. Required mostly in situation where there are multiple smoke tests in a package and each requires different paramteres set
+`<param_filename>` - [optional] param filename. Default value is `test.param.yaml`. Required mostly in situation where there are multiple smoke tests in a package and each requires different parameters set.
 
 `<arguments>` - [optional] arguments passed to executable. By default no arguments are passed.
 
 which adds `<executable_name>_smoke_test` test to suite.
 
 Example test result:
 
-```
+```log
 build/<package_name>/test_results/<package_name>/<executable_name>_smoke_test.xunit.xml: 1 test, 0 errors, 0 failures, 0 skipped
 
 ```
 
+### Interface test
+
+To add a interface test to your package tests, add test dependency on `autoware_testing` to `package.xml`
+
+```xml
+<test_depend>autoware_testing</test_depend>
+```
+
+and add a yaml file that defines topics to your package. The file will looks like this:
+
+```yaml
+input_topics:
+  - autoware_auto_control_msgs/msg/AckermannLateralCommand
+  - autoware_auto_control_msgs/msg/LongitudinalCommand
+output_topics:
+  - autoware_auto_control_msgs/msg/AckermannControlCommand
+```
```

Comment:
Don't you need topic names here?
If not, how can the test library know that?

Reply:
The topic name will change in the remap, so a Launch file will be needed.
At this stage, we are only concerned with the topic type to simplify matters.

Reply:
Hmm? What happens if a node subscribes to two topics of the same type?

Reply:
It seems that the design intent has not been conveyed very well. I will be updating the design document to show you a more detailed design. Please wait for a while.

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/391/files#diff-a790f3b1a2981da08ef8d466c4e13a5086a78ddba58a9f88f8b3b1b1c877525dR127

I have reflected the comments. Changed to validate the name and type given in the code.
I was thinking remap should be considered, but that is out of scope.